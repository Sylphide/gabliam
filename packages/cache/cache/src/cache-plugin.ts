import { Scan, Plugin, GabliamPlugin } from '@gabliam/core';

@Plugin()
@Scan()
export class CachePlugin implements GabliamPlugin {}
