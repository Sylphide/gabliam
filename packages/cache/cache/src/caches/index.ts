export * from './memory-cache';
export * from './no-op-cache';
