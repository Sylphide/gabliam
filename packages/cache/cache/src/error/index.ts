export * from './cache-manager-pgk-not-installed-error';
export * from './cache-name-is-mandatory';
export * from './cache-pgk-not-installed-error';
