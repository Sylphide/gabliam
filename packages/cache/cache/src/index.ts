export * from './cache';
export * from './cache-manager';
export { CachePlugin as default } from './cache-plugin';
export * from './caches';
export * from './constant';
export * from './metadatas';
export * from './simple-cache-manager';
