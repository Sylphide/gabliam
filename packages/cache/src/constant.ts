export const CACHE_MANAGER = Symbol('CACHE_MANAGER');

export const NO_RESULT = Symbol('NO_RESULT');

export const METADATA_KEY = {
  cacheGroup: '_cacheGroup',
};

export const ERRORS_MSGS = {
  DUPLICATED_CACHE_DECORATOR: `Cannot apply @Cache decorator multiple times.`,
};
