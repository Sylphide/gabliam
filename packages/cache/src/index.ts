export * from './caches';
export * from './cache-manager';
export * from './cache';
export * from './simple-cache-manager';
