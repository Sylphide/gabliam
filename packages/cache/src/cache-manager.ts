import { Cache, ConstructableCache } from './cache';

/**
 * Gabliam central cache manager SPI.
 * Allows for retrieving named {@link Cache} regions.
 */
export interface CacheManager {
  /**
   * Return the cache associated with the given name.
   * @param name the cache identifier (must not be {@code null})
   * @return the associated cache, or {@code undefined} if none found
   */
  getCache(name: string): Promise<Cache | undefined>;

  /**
   * Return a collection of the cache names known by this manager.
   * @return the names of all caches known by the cache manager
   */
  getCacheNames(): string[];
}

export interface ConstructableCacheManager {
  new (
    cacheMap: Map<string, Cache>,
    dynamic: boolean,
    defaultCache: ConstructableCache,
    defaultOptionsCache?: object
  ): CacheManager;
}
