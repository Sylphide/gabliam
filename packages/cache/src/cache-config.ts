import { Bean, InjectContainer, PluginConfig, Value, Joi } from '@gabliam/core';
import { CACHE_MANAGER } from './constant';
import { ConstructableCacheManager } from './cache-manager';
import { ConstructableCache, Cache } from './cache';
import * as d from 'debug';
import { SimpleCacheManager } from './simple-cache-manager';
import {
  CachePgkNotInstalledError,
  CacheManagerPgkNotInstalledError
} from './error';
import { MemoryCache, NoOpCache } from './caches';
const debug = d('Gabliam:Plugin:CachePlugin');

export interface PluginConfig {
  cacheManager: string | ConstructableCacheManager;

  cacheMap: { [k: string]: PluginCacheConfig } | undefined;

  dynamic: boolean;

  defaultCache: string | ConstructableCache;

  defaultOptionsCache?: Object;
}

export interface PluginCacheConfig {
  cache: string | ConstructableCache;

  options?: Object;
}

const stringOrClass = Joi.alternatives().try([Joi.string(), Joi.func()]);

const pluginValidator = Joi.object().keys({
  cacheManager: stringOrClass.default('SimpleCacheManager'),
  dynamic: Joi.boolean().default(true),
  cacheMap: Joi.object()
    .pattern(
      /.*/,
      Joi.object({
        cache: stringOrClass.required(),
        options: Joi.object()
      })
    )
    .default(undefined),
  defaultCache: stringOrClass.default('NoOpCache'),
  defaultOptionsCache: Joi.object()
});

@InjectContainer()
@PluginConfig()
export class CachePluginConfig {
  @Value('application.cacheConfig', pluginValidator)
  cacheConfig: PluginConfig | undefined;

  @Bean(CACHE_MANAGER)
  createCacheManager(): any {
    if (this.cacheConfig === undefined) {
      return this.createDefaultManager();
    }
    const cacheMap = new Map<string, Cache>();
    if (this.cacheConfig.cacheMap) {
      for (const [cacheKey, { cache, options }] of Object.entries(
        this.cacheConfig.cacheMap
      )) {
        const CacheConstruc = this.getCacheConstruct(cache);
        cacheMap.set(cacheKey, new CacheConstruc(cacheKey, options));
      }
    }
    const CacheManagerConstruct = this.getCacheManagerConstruct(
      this.cacheConfig.cacheManager
    );

    return new CacheManagerConstruct(
      cacheMap,
      this.cacheConfig.dynamic,
      this.getCacheConstruct(this.cacheConfig.defaultCache),
      this.cacheConfig.defaultOptionsCache
    );
  }

  private getCacheManagerConstruct(
    cacheManager: string | ConstructableCacheManager
  ): ConstructableCacheManager {
    if (typeof cacheManager === 'string') {
      switch (cacheManager) {
        case 'SimpleCacheManager':
          return SimpleCacheManager;
        default:
          try {
            return <ConstructableCacheManager>require(cacheManager).default;
          } catch {
            throw new CacheManagerPgkNotInstalledError(cacheManager);
          }
      }
    }
    return cacheManager;
  }

  private getCacheConstruct(
    cache: string | ConstructableCache
  ): ConstructableCache {
    if (typeof cache === 'string') {
      switch (cache) {
        case 'MemoryCache':
          return MemoryCache;
        case 'NoOpCache':
          return NoOpCache;
        default:
          try {
            return <ConstructableCache>require(cache).default;
          } catch {
            throw new CachePgkNotInstalledError(cache);
          }
      }
    }

    return cache;
  }

  private createDefaultManager() {
    debug('Create Defaut cache Manager');
    return new SimpleCacheManager(new Map<string, Cache>(), true);
  }
}
