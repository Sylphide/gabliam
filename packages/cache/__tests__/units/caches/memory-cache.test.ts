import { MemoryCache } from '../../../src/index';

let cache: MemoryCache;

beforeEach(() => {
  cache = new MemoryCache('test');
});

test('cache', () => {
  expect(cache.getName()).toMatchSnapshot();
  expect(cache.getNativeCache()).toMatchSnapshot();
});

test('cache get & put', () => {
  expect(cache.get('test')).toMatchSnapshot();
  cache.put('test', 'test');
  expect(cache).toMatchSnapshot();
  expect(cache.get('test')).toMatchSnapshot();
});

test('cache putIfAbsent', () => {
  cache.put('test', 'test');
  cache.putIfAbsent('test', 'testnew');
  cache.putIfAbsent('test2', 'test2');
  cache.putIfAbsent('test3', 'test3');
  expect(cache).toMatchSnapshot();
  expect(cache.get('test')).toMatchSnapshot();
  expect(cache.get('test2')).toMatchSnapshot();
});

test('evict', () => {
  cache.putIfAbsent('test', 'testnew');
  cache.putIfAbsent('test2', 'test2');
  cache.putIfAbsent('test3', 'test3');
  expect(cache).toMatchSnapshot();
  cache.evict('test3');
});

test('clear', () => {
  cache.putIfAbsent('test', 'testnew');
  cache.putIfAbsent('test2', 'test2');
  cache.putIfAbsent('test3', 'test3');
  expect(cache).toMatchSnapshot();
  cache.clear();
  expect(cache).toMatchSnapshot();
});
