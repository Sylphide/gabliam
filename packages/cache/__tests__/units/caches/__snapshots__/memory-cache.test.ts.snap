// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cache 1`] = `"test"`;

exports[`cache 2`] = `
MemoryCache {
  "cache": LRUCache {
    Symbol(max): Infinity,
    Symbol(lengthCalculator): [Function],
    Symbol(allowStale): false,
    Symbol(maxAge): 0,
    Symbol(dispose): undefined,
    Symbol(noDisposeOnSet): false,
    Symbol(cache): Map {},
    Symbol(lruList): Yallist {
      "head": null,
      "length": 0,
      "tail": null,
    },
    Symbol(length): 0,
  },
  "name": "test",
}
`;

exports[`cache get & put 1`] = `undefined`;

exports[`cache get & put 2`] = `
MemoryCache {
  "cache": LRUCache {
    Symbol(max): Infinity,
    Symbol(lengthCalculator): [Function],
    Symbol(allowStale): false,
    Symbol(maxAge): 0,
    Symbol(dispose): undefined,
    Symbol(noDisposeOnSet): false,
    Symbol(cache): Map {
      "test" => Node {
        "list": Yallist {
          "head": [Circular],
          "length": 1,
          "tail": [Circular],
        },
        "next": null,
        "prev": null,
        "value": Entry {
          "key": "test",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test",
        },
      },
    },
    Symbol(lruList): Yallist {
      "head": Node {
        "list": [Circular],
        "next": null,
        "prev": null,
        "value": Entry {
          "key": "test",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test",
        },
      },
      "length": 1,
      "tail": Node {
        "list": [Circular],
        "next": null,
        "prev": null,
        "value": Entry {
          "key": "test",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test",
        },
      },
    },
    Symbol(length): 1,
  },
  "name": "test",
}
`;

exports[`cache get & put 3`] = `"test"`;

exports[`cache putIfAbsent 1`] = `
MemoryCache {
  "cache": LRUCache {
    Symbol(max): Infinity,
    Symbol(lengthCalculator): [Function],
    Symbol(allowStale): false,
    Symbol(maxAge): 0,
    Symbol(dispose): undefined,
    Symbol(noDisposeOnSet): false,
    Symbol(cache): Map {
      "test" => Node {
        "list": Yallist {
          "head": Node {
            "list": [Circular],
            "next": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": [Circular],
              "value": Entry {
                "key": "test2",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test2",
              },
            },
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "length": 3,
          "tail": [Circular],
        },
        "next": null,
        "prev": Node {
          "list": Yallist {
            "head": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": null,
              "value": Entry {
                "key": "test3",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test3",
              },
            },
            "length": 3,
            "tail": [Circular],
          },
          "next": [Circular],
          "prev": Node {
            "list": Yallist {
              "head": [Circular],
              "length": 3,
              "tail": [Circular],
            },
            "next": [Circular],
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "value": Entry {
          "key": "test",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test",
        },
      },
      "test2" => Node {
        "list": Yallist {
          "head": Node {
            "list": [Circular],
            "next": [Circular],
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "length": 3,
          "tail": Node {
            "list": [Circular],
            "next": null,
            "prev": [Circular],
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test",
            },
          },
        },
        "next": Node {
          "list": Yallist {
            "head": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": null,
              "value": Entry {
                "key": "test3",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test3",
              },
            },
            "length": 3,
            "tail": [Circular],
          },
          "next": null,
          "prev": [Circular],
          "value": Entry {
            "key": "test",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test",
          },
        },
        "prev": Node {
          "list": Yallist {
            "head": [Circular],
            "length": 3,
            "tail": Node {
              "list": [Circular],
              "next": null,
              "prev": [Circular],
              "value": Entry {
                "key": "test",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test",
              },
            },
          },
          "next": [Circular],
          "prev": null,
          "value": Entry {
            "key": "test3",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test3",
          },
        },
        "value": Entry {
          "key": "test2",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test2",
        },
      },
      "test3" => Node {
        "list": Yallist {
          "head": [Circular],
          "length": 3,
          "tail": Node {
            "list": [Circular],
            "next": null,
            "prev": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": [Circular],
              "value": Entry {
                "key": "test2",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test2",
              },
            },
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test",
            },
          },
        },
        "next": Node {
          "list": Yallist {
            "head": [Circular],
            "length": 3,
            "tail": Node {
              "list": [Circular],
              "next": null,
              "prev": [Circular],
              "value": Entry {
                "key": "test",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test",
              },
            },
          },
          "next": Node {
            "list": Yallist {
              "head": [Circular],
              "length": 3,
              "tail": [Circular],
            },
            "next": null,
            "prev": [Circular],
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test",
            },
          },
          "prev": [Circular],
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "prev": null,
        "value": Entry {
          "key": "test3",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test3",
        },
      },
    },
    Symbol(lruList): Yallist {
      "head": Node {
        "list": [Circular],
        "next": Node {
          "list": [Circular],
          "next": Node {
            "list": [Circular],
            "next": null,
            "prev": [Circular],
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test",
            },
          },
          "prev": [Circular],
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "prev": null,
        "value": Entry {
          "key": "test3",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test3",
        },
      },
      "length": 3,
      "tail": Node {
        "list": [Circular],
        "next": null,
        "prev": Node {
          "list": [Circular],
          "next": [Circular],
          "prev": Node {
            "list": [Circular],
            "next": [Circular],
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "value": Entry {
          "key": "test",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test",
        },
      },
    },
    Symbol(length): 3,
  },
  "name": "test",
}
`;

exports[`cache putIfAbsent 2`] = `"test"`;

exports[`cache putIfAbsent 3`] = `"test2"`;

exports[`clear 1`] = `
MemoryCache {
  "cache": LRUCache {
    Symbol(max): Infinity,
    Symbol(lengthCalculator): [Function],
    Symbol(allowStale): false,
    Symbol(maxAge): 0,
    Symbol(dispose): undefined,
    Symbol(noDisposeOnSet): false,
    Symbol(cache): Map {
      "test" => Node {
        "list": Yallist {
          "head": Node {
            "list": [Circular],
            "next": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": [Circular],
              "value": Entry {
                "key": "test2",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test2",
              },
            },
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "length": 3,
          "tail": [Circular],
        },
        "next": null,
        "prev": Node {
          "list": Yallist {
            "head": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": null,
              "value": Entry {
                "key": "test3",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test3",
              },
            },
            "length": 3,
            "tail": [Circular],
          },
          "next": [Circular],
          "prev": Node {
            "list": Yallist {
              "head": [Circular],
              "length": 3,
              "tail": [Circular],
            },
            "next": [Circular],
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "value": Entry {
          "key": "test",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "testnew",
        },
      },
      "test2" => Node {
        "list": Yallist {
          "head": Node {
            "list": [Circular],
            "next": [Circular],
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "length": 3,
          "tail": Node {
            "list": [Circular],
            "next": null,
            "prev": [Circular],
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "testnew",
            },
          },
        },
        "next": Node {
          "list": Yallist {
            "head": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": null,
              "value": Entry {
                "key": "test3",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test3",
              },
            },
            "length": 3,
            "tail": [Circular],
          },
          "next": null,
          "prev": [Circular],
          "value": Entry {
            "key": "test",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "testnew",
          },
        },
        "prev": Node {
          "list": Yallist {
            "head": [Circular],
            "length": 3,
            "tail": Node {
              "list": [Circular],
              "next": null,
              "prev": [Circular],
              "value": Entry {
                "key": "test",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "testnew",
              },
            },
          },
          "next": [Circular],
          "prev": null,
          "value": Entry {
            "key": "test3",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test3",
          },
        },
        "value": Entry {
          "key": "test2",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test2",
        },
      },
      "test3" => Node {
        "list": Yallist {
          "head": [Circular],
          "length": 3,
          "tail": Node {
            "list": [Circular],
            "next": null,
            "prev": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": [Circular],
              "value": Entry {
                "key": "test2",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test2",
              },
            },
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "testnew",
            },
          },
        },
        "next": Node {
          "list": Yallist {
            "head": [Circular],
            "length": 3,
            "tail": Node {
              "list": [Circular],
              "next": null,
              "prev": [Circular],
              "value": Entry {
                "key": "test",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "testnew",
              },
            },
          },
          "next": Node {
            "list": Yallist {
              "head": [Circular],
              "length": 3,
              "tail": [Circular],
            },
            "next": null,
            "prev": [Circular],
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "testnew",
            },
          },
          "prev": [Circular],
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "prev": null,
        "value": Entry {
          "key": "test3",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test3",
        },
      },
    },
    Symbol(lruList): Yallist {
      "head": Node {
        "list": [Circular],
        "next": Node {
          "list": [Circular],
          "next": Node {
            "list": [Circular],
            "next": null,
            "prev": [Circular],
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "testnew",
            },
          },
          "prev": [Circular],
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "prev": null,
        "value": Entry {
          "key": "test3",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test3",
        },
      },
      "length": 3,
      "tail": Node {
        "list": [Circular],
        "next": null,
        "prev": Node {
          "list": [Circular],
          "next": [Circular],
          "prev": Node {
            "list": [Circular],
            "next": [Circular],
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "value": Entry {
          "key": "test",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "testnew",
        },
      },
    },
    Symbol(length): 3,
  },
  "name": "test",
}
`;

exports[`clear 2`] = `
MemoryCache {
  "cache": LRUCache {
    Symbol(max): Infinity,
    Symbol(lengthCalculator): [Function],
    Symbol(allowStale): false,
    Symbol(maxAge): 0,
    Symbol(dispose): undefined,
    Symbol(noDisposeOnSet): false,
    Symbol(cache): Map {},
    Symbol(lruList): Yallist {
      "head": null,
      "length": 0,
      "tail": null,
    },
    Symbol(length): 0,
  },
  "name": "test",
}
`;

exports[`evict 1`] = `
MemoryCache {
  "cache": LRUCache {
    Symbol(max): Infinity,
    Symbol(lengthCalculator): [Function],
    Symbol(allowStale): false,
    Symbol(maxAge): 0,
    Symbol(dispose): undefined,
    Symbol(noDisposeOnSet): false,
    Symbol(cache): Map {
      "test" => Node {
        "list": Yallist {
          "head": Node {
            "list": [Circular],
            "next": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": [Circular],
              "value": Entry {
                "key": "test2",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test2",
              },
            },
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "length": 3,
          "tail": [Circular],
        },
        "next": null,
        "prev": Node {
          "list": Yallist {
            "head": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": null,
              "value": Entry {
                "key": "test3",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test3",
              },
            },
            "length": 3,
            "tail": [Circular],
          },
          "next": [Circular],
          "prev": Node {
            "list": Yallist {
              "head": [Circular],
              "length": 3,
              "tail": [Circular],
            },
            "next": [Circular],
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "value": Entry {
          "key": "test",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "testnew",
        },
      },
      "test2" => Node {
        "list": Yallist {
          "head": Node {
            "list": [Circular],
            "next": [Circular],
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "length": 3,
          "tail": Node {
            "list": [Circular],
            "next": null,
            "prev": [Circular],
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "testnew",
            },
          },
        },
        "next": Node {
          "list": Yallist {
            "head": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": null,
              "value": Entry {
                "key": "test3",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test3",
              },
            },
            "length": 3,
            "tail": [Circular],
          },
          "next": null,
          "prev": [Circular],
          "value": Entry {
            "key": "test",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "testnew",
          },
        },
        "prev": Node {
          "list": Yallist {
            "head": [Circular],
            "length": 3,
            "tail": Node {
              "list": [Circular],
              "next": null,
              "prev": [Circular],
              "value": Entry {
                "key": "test",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "testnew",
              },
            },
          },
          "next": [Circular],
          "prev": null,
          "value": Entry {
            "key": "test3",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test3",
          },
        },
        "value": Entry {
          "key": "test2",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test2",
        },
      },
      "test3" => Node {
        "list": Yallist {
          "head": [Circular],
          "length": 3,
          "tail": Node {
            "list": [Circular],
            "next": null,
            "prev": Node {
              "list": [Circular],
              "next": [Circular],
              "prev": [Circular],
              "value": Entry {
                "key": "test2",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "test2",
              },
            },
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "testnew",
            },
          },
        },
        "next": Node {
          "list": Yallist {
            "head": [Circular],
            "length": 3,
            "tail": Node {
              "list": [Circular],
              "next": null,
              "prev": [Circular],
              "value": Entry {
                "key": "test",
                "length": 1,
                "maxAge": 0,
                "now": 0,
                "value": "testnew",
              },
            },
          },
          "next": Node {
            "list": Yallist {
              "head": [Circular],
              "length": 3,
              "tail": [Circular],
            },
            "next": null,
            "prev": [Circular],
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "testnew",
            },
          },
          "prev": [Circular],
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "prev": null,
        "value": Entry {
          "key": "test3",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test3",
        },
      },
    },
    Symbol(lruList): Yallist {
      "head": Node {
        "list": [Circular],
        "next": Node {
          "list": [Circular],
          "next": Node {
            "list": [Circular],
            "next": null,
            "prev": [Circular],
            "value": Entry {
              "key": "test",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "testnew",
            },
          },
          "prev": [Circular],
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "prev": null,
        "value": Entry {
          "key": "test3",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "test3",
        },
      },
      "length": 3,
      "tail": Node {
        "list": [Circular],
        "next": null,
        "prev": Node {
          "list": [Circular],
          "next": [Circular],
          "prev": Node {
            "list": [Circular],
            "next": [Circular],
            "prev": null,
            "value": Entry {
              "key": "test3",
              "length": 1,
              "maxAge": 0,
              "now": 0,
              "value": "test3",
            },
          },
          "value": Entry {
            "key": "test2",
            "length": 1,
            "maxAge": 0,
            "now": 0,
            "value": "test2",
          },
        },
        "value": Entry {
          "key": "test",
          "length": 1,
          "maxAge": 0,
          "now": 0,
          "value": "testnew",
        },
      },
    },
    Symbol(length): 3,
  },
  "name": "test",
}
`;
